File: finanze/infrastructure/client/rates/metal/gold_api_price_client.py
import logging
from typing import Optional

import httpx
from domain.commodity import COMMODITY_SYMBOLS, CommodityType, WeightUnit
from domain.dezimal import Dezimal
from domain.exchange_rate import CommodityExchangeRate
from infrastructure.client.http.http_session import get_http_session


class GoldApiPriceClient:
    BASE_URL = "https://api.gold-api.com/price"
    TIMEOUT = 3

    SUPPORTED_COMMODITIES = {
        CommodityType.GOLD,
        CommodityType.SILVER,
        CommodityType.PALLADIUM,
    }

    def __init__(self):
        self._log = logging.getLogger(__name__)
        self._session = get_http_session()

    async def get_price(
        self, commodity: CommodityType, timeout: int | None = None
    ) -> Optional[CommodityExchangeRate]:
        if commodity not in self.SUPPORTED_COMMODITIES:
            raise ValueError(f"Unsupported commodity type: {commodity}")

        return await self._fetch_price(
            COMMODITY_SYMBOLS.get(commodity).upper(), timeout or self.TIMEOUT
        )

    async def _fetch_price(
        self, symbol: str, timeout: int
    ) -> Optional[CommodityExchangeRate]:
        url = f"{self.BASE_URL}/{symbol}"
        try:
            data = await self._fetch(url, timeout)
        except (httpx.RequestError, TimeoutError) as e:
            self._log.error(f"Timeout fetching price for {symbol}: {e}")
            return None

        return CommodityExchangeRate(
            unit=WeightUnit.TROY_OUNCE,
            currency="USD",
            price=Dezimal(str(data["price"])),
        )

    async def _fetch(self, url: str, request_timeout: int) -> dict:
        response = await self._session.get(url, timeout=request_timeout)
        if response.ok:
            return await response.json()

        body = await response.text()
        self._log.error("Error Response Body:" + body)
        response.raise_for_status()
        return {}

File: mobile/src/services/client/rates/metal/goldApiPriceClient.ts
import { CommodityExchangeRate, CommodityType, WeightUnit } from "@/domain"
import { Dezimal } from "@/domain/dezimal"
import { COMMODITY_SYMBOLS } from "@/domain/constants/commodity"
import { httpGetJson } from "@/services/client/http/httpClient"

export class GoldApiPriceClient {
  static readonly BASE_URL = "https://api.gold-api.com/price"
  static readonly TIMEOUT = 3

  static readonly SUPPORTED_COMMODITIES = new Set<CommodityType>([
    CommodityType.GOLD,
    CommodityType.SILVER,
    CommodityType.PALLADIUM,
  ])

  async getPrice(
    commodity: CommodityType,
    timeout?: number | null,
  ): Promise<CommodityExchangeRate | null> {
    if (!GoldApiPriceClient.SUPPORTED_COMMODITIES.has(commodity)) {
      throw new Error(`Unsupported commodity type: ${commodity}`)
    }

    const symbol = COMMODITY_SYMBOLS[commodity].toUpperCase()
    return this.fetchPrice(symbol, timeout ?? GoldApiPriceClient.TIMEOUT)
  }

  private async fetchPrice(
    symbol: string,
    timeoutSec: number,
  ): Promise<CommodityExchangeRate | null> {
    const url = `${GoldApiPriceClient.BASE_URL}/${symbol}`

    try {
      // Use JSON helper to stay aligned with backend "requests.get().json()".
      const { response, data } = await httpGetJson<any>(
        url,
        undefined,
        timeoutSec,
      )
      if (!response.ok) {
        console.error("Error Response Body:" + JSON.stringify(data))
        throw new Error(`Gold API request failed: ${response.status}`)
      }

      return {
        unit: WeightUnit.TROY_OUNCE,
        currency: "USD",
        price: Dezimal.fromString(String((data as any).price)),
      }
    } catch (e: any) {
      // Backend returns None on Timeout only; our timeout throws AbortError.
      if (e?.name === "HttpTimeoutError" || e?.name === "AbortError") {
        console.error(`Timeout fetching price for ${symbol}:`, e)
        return null
      }
      throw e
    }
  }
}
